### ISSUE TITLE
StackNavigator as formSheet content has size of zero (iOS)

### ISSUE THREAD

**Issue by arnekolja** [software-mansion/react-native-screens#2522]
### Description

After upgrading to react-navigation 7 and react-native-screens 4.2.0 my modal's content, a native stack navigator, isn't showing up any more. After some debugging I found out that it has a layout of zero width and zero height. Applying a wrapping `<View />` with fixed width and height works, applying `flex: 1` does not. However, a fixed height cannot be the solution for my flexible-height content. Using a normal view works fine and grows with its content, but not with the stack navigator.

This does only apply to `presentation: 'formSheet'` but not `presentation: 'modal'`, despite the fact that at least on iOS (only tested on iOS) the resulting modal looks the same. `presentation: 'modal'` handles sizing of my stack navigator correctly.

1. Is `presentation: 'formSheet'` and `presentation: 'modal'` really the same UI element here? And how can I fix the layout/height of my content? It is definitely a change from 3.x to 4.x, as it worked on 3.x for a long time.

### Steps to reproduce

Place a native stack navigator inside a modal of type 'formSheet'.

### Snack or a link to a repository

https://snack.expo.dev/Pqu3Cikm2Pk84jev-qtF9

### Screens version

4.2.0

### React Native version

0.76.2

### Platforms

iOS

### JavaScript runtime

Hermes

### Workflow

React Native (without Expo)

### Architecture

Fabric (New Architecture)

### Build type

Debug mode

### Device

iOS simulator

### Device model

iPhone 15 (iOS 17.2)

### Acknowledgements

Yes

**Comment by colin-chadwick** [comment]
Having the same issue.

**Comment by bojandurmic** [comment]
Same here `flex: 1` just won't work, I ended up using fixed height, but definitely not happy with that solution.

**Comment by kkafar** [comment]
Are you using `formSheet` for its ability to specify detents or is `modal` presentation good replacement enough? 

The `formSheet` presentation has changed with 4.0 (see [here](https://github.com/software-mansion/react-native-screens/releases/tag/4.0.0)) and it is a bit impaired (but has other possibilities) at the moment. I'll try to handle this asap

**Comment by colin-chadwick** [comment]
@kkafar Yes, the formSheet allows for better customization than the modal (for e.g. specifying detents, showing the grabber, corner radius, etc.).

**Comment by kkafar** [comment]
Then please make sure to read the [release notes](https://github.com/software-mansion/react-native-screens/releases/tag/4.0.0)

**Comment by arnekolja** [comment]
For my two use cases it is okay to fallback to 'modal'. I'm interested in the solution tho. But I'll keep a view on this ticket and re-read the release notes etc. Thanks for the response 🙌🏾

**Comment by kkafar** [comment]
I can't give promises on timelines but will prioritise fixing this up

**Comment by bojandurmic** [comment]
Thank you @kkafar for your effort and communication. We really appreciate it!

**Comment by jgo80** [comment]
Any news? I'm using modal as fallback right now, but here the modal strangely "deforms" while sliding up in iOS, messing up all the content, until it's settled. never had that with formSheet.

**Comment by snndmnsz** [comment]
This issue still continues.

![Image](https://github.com/user-attachments/assets/4b8e3682-115b-4cde-96be-beabefb914c0)

**Comment by hraschan** [comment]
Are there any updates on this?

**Comment by damiensedgwick** [comment]
This seems to be working for me on iOS but I am experience the issue with Android simulators.

**Comment by snndmnsz** [comment]
Giving `height: '100%'` to contentStyle in your parent stack screen is seems to working. I called my collection stack form bottom tab stack. Its working even if they are not in same stack, you can call form another stack and it still formSheet style.  I tried giving `flex:1` and  did not worked still a black screen. Just give 100% height

```js
<Stack.Screen
  name="collection"
  options={{
    presentation: "formSheet",
    contentStyle: {
      height: '100%',
    },
    sheetCornerRadius: 20,
  }}
/>
```
![Image](https://github.com/user-attachments/assets/56332625-c14c-4868-9b60-d6ded64726fc)

**Comment by danieljvdm** [comment]
> Giving `height: '100%'` to contentStyle in your parent stack screen seems to work. I called my collection stack from the bottom tab stack. It works even if they are not in the same stack, you can call it from another stack, and it will still bring screen as formSheet. I tried giving flex: 1, but it did not work, it was still a blank screen. Just give it height: '100%'.
> 
> <Stack.Screen
>       name="collection"
>       options={{
>           presentation: "formSheet",
>           contentStyle:{
>             height: '100%',
>           },
>           sheetCornerRadius: 20,
>           sheetGrabberVisible: true,
>        }}
>    />
> ![Image](https://github.com/user-attachments/assets/c304dfc9-d802-4d72-88fa-de48a4320a49)

This is a hacky fix at best, if you add a background color the sheet you'll see the animations are very laggy, you can just get away with it if your background doesn't change. If you have content fixed to the bottom it won't look good.

**Comment by Patrick-web** [comment]
> Giving `height: '100%'` to contentStyle in your parent stack screen seems to work. I called my collection stack from the bottom tab stack. It works even if they are not in the same stack, you can call it from another stack, and it will still bring screen as formSheet. I tried giving flex: 1, but it did not work, it was still a blank screen. Just give it height: '100%'.
> 
> <Stack.Screen
>       name="collection"
>       options={{
>           presentation: "formSheet",
>           contentStyle:{
>             height: '100%',
>           },
>           sheetCornerRadius: 20,
>           sheetGrabberVisible: true,
>        }}
>    />
> ![Image](https://github.com/user-attachments/assets/c304dfc9-d802-4d72-88fa-de48a4320a49)

This fix should be built in

**Comment by kylealwyn** [comment]
Hi, just reporting in that this still seems to be an issue on 4.11.1 and setting a contentStyle height of 100% does not play nicely with `fitToContents` when trying to implement small, navigatable popover sheets

**Comment by galgord** [comment]
Does anyone have any solution for this I am facing the same issue.

**Comment by galgord** [comment]
@kkafar Do you think it would be possible to take another look at this? I am happt to help reproduce if needed.

**Comment by sumersao** [comment]
I'm also facing the same issue!

**Comment by Bowlerr** [comment]
I believe I have found the issue....

After a stupid amount of time debugging this issue in our project. I found out if you remove the background colour from the main container view in the form-sheets page. It just works...

@sumersao  @galgord

**Comment by jgo80** [comment]
> I believe I have found the issue....
> 
> After a stupid amount of time debugging this issue in our project. I found out if you remove the background colour from the main container view in the form-sheets page. It just works...
> 
> [@sumersao](https://github.com/sumersao) [@galgord](https://github.com/galgord)

I cannot confirm this. Neither my stacks nor their wrappers have a background color.

**Comment by Bowlerr** [comment]
The page itself that is being loaded?

**Comment by Bowlerr** [comment]
@jgo80

**Comment by Bowlerr** [comment]
Like on that component itself.

**Comment by Bowlerr** [comment]
Setting a parent view with just flex 1 and nothing else solved this. It was odd but it was the background colour causing the issue for me

**Comment by jgo80** [comment]
Nope. Ally my parent views of the modal content are `<View {...props} style={{flex:1}} />`

**Comment by Bowlerr** [comment]
Ah never mind, I hope you find the cause. The only other thing I'm doing is setting the background colour in the contentStyle for the sheet.

Thought I would share in case it solves it for y'all

**Comment by kylealwyn** [comment]
@kkafar this is blocking some nice, contextual experiences. hoping you can take another look. happy to pair. thank you.

**Comment by anckaertv** [comment]
Any update on this one ? 👍

**Comment by alextbogdanov** [comment]
+ 1

**Comment by jgo80** [comment]
Seems to work now with react-native-screens@~4.18.0

**Comment by vladimir-vucetic** [comment]
> Seems to work now with react-native-screens@~4.18.0

It fix this issue but liquid glass tab bar icons are not visible if update to this version on expo 54.
`<NativeTabs.Trigger name="home">
          <Label>Home</Label>
          <Icon sf="house.fill" drawable="custom_android_drawable" />
        </NativeTabs.Trigger>
`
<img width="335" height="108" alt="Image" src="https://github.com/user-attachments/assets/c53d9680-a605-4811-9fec-bd51e51276ee" />

**Comment by PierreCapo** [comment]
To me it seems `4.18.0` still has the issue

**Comment by amana-othman** [comment]
> To me it seems `4.18.0` still has the issue

It is fixed in 4.18.0, check the package version in yarn.lock.

**Comment by t0maboro** [comment]
This requires some additional explanation.

During my previous updates made to the `FormSheet` component related to iOS 26 - specifically with enabling `flex` inside the content, I reintroduced some kind of regression that caused flickering or jumping content. This was due to asynchronous state updates in the custom shadow node for the `Screen` component.

As a result, right after releasing the version `4.18.0`, the initial fix was reverted: https://github.com/software-mansion/react-native-screens/commit/460569215747281440102693f7759795aefd3cdf

However, starting from React Native 0.82, Meta has introduced a new API that allows for **synchronous** state updates. This resolves the flickering issue caused by **asynchronous** updates.

Therefore, I'm introducing a PR https://github.com/software-mansion/react-native-screens/pull/3454 , that allows dynamic sizing under the following conditions:

- The FormSheet is not using `sheetDetents: "fitToContents"`, because in this configuration, the sheet size is determined by the content itself, and flex-based dynamic sizing cannot work reliably.
- The React Native version is 0.82 or higher, as this is when the new synchronous state update API becomes available.
- The feature flag `featureFlags.experiment.synchronousScreenUpdatesEnabled` from `react-native-screens` is enabled.

I hope this solution is acceptable and helps us avoid known regressions that, unfortunately, depend on the specific RN version being used in this case.

**Comment by julien9999** [comment]
what about on android?

**Comment by t0maboro** [comment]
Hi, 
I haven't noticed any issues on Android so far. On Android, the size of the FormSheet is determined based on:
- its maximum detent, when we're using fractions, 
- or, in the case of fitToContents, it's based on the content size.

If your question refers to using `fitToContents` in combination with `flex: 1` on the top-level container inside the content, it's a known technical limitation, because the FormSheet doesn't have any container which could be treated as a reference for determining its size. ref. https://reactnavigation.org/docs/native-stack-navigator#using-form-sheet

<img width="714" height="831" alt="Image" src="https://github.com/user-attachments/assets/fe9d4d32-8816-4d29-b6c8-34d72d3a58e8" />

However, if you have a reproducible case on Android that doesn't fall under the scenarios mentioned above - feel free to report it as a separate issue.

**Comment by Lamperoyge** [comment]
thank you so much! couldn't figure out what's wrong with the formSheet haha

**Comment by DavoCg** [comment]
Hello everyone, sorry to re open this closed issue. 

@t0maboro with latest version on `react-native-screens` (4.19.0) and RN 82+ it seems that indeed the `flex-1` is now working in `formSheet` but do we agree that for now it is still impossible to have a nested stack navigator inside a formSheet that would fitToContent ? 

Best
